### PR TITLE
Update triggers default configmap for runAsUser and runAsGroup to handle restricted securityContext for Triggers

### PR DESCRIPTION
**Pull request by savitaashture**
Context:

As part of addressing https://issues.redhat.com/browse/OCPSTRAT-487, there's a plan to enable restricted security context by default starting from Openshift 4.17. 
Once this becomes the default setting, existing Triggers functionality may break. 
This is because we currently set security context to false, and the pipelines-scc security context constraint (SCC) doesn't have seccompProfiles: runtime/default, which is required when restricted security context is enabled by default because by default triggers set `runAsUser and `runAsGroup` to 65532.

So Updated Triggers only to accept values for `runAsUser and `runAsGroup` through CM and using Operator we are setting to `""` so that Openshift can set those random values.

Tested below scenarios:

1 set security context to restricted for a namespace  (`oc label ns test pod-security.kubernetes.io/enforce=restricted --overwrite=true`)
  * No seccompProfiles: runtime/default to `pipelines-scc`
 and Created EL, POD is up and running
     ```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1000730000
      seccompProfile:
        type: RuntimeDefault

     ```
      
      
2 remove security context to restricted for a namespace  (`oc label ns test pod-security.kubernetes.io/enforce-`)
  * No seccompProfiles: runtime/default to `pipelines-scc`
 and Created EL, POD is up and running
     ```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1000710000
      seccompProfile:
        type: RuntimeDefault
     ```
	

Fixes : https://issues.redhat.com/browse/SRVKP-4372

Signed-off-by: Savita Ashture <sashture@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Run `make test lint` before submitting a PR
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

**Comment by savitaashture** [comment]
@tektoncd/operator-maintainers Will bring this issue in WG call also wanted to talk `set-security-context` related to Pipeline

**Comment by savitaashture** [comment]
@piyush-garg 
I have verified the question which you bought during WG call

When `el-security-context` is true
EL pod have below securityContext
```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
        - MKNOD
      runAsGroup: 65532
      runAsNonRoot: true
      runAsUser: 65532
      seccompProfile:
        type: RuntimeDefault
    terminationMessagePath: /dev/termination-log
```

And pod is up and running with those securityContext values
```
$ kubectl get pods 
NAME                                  READY   STATUS      RESTARTS   AGE
el-github-listener-7ff9675c4d-jpctb   1/1     Running     0          41s
github-run-fhggv-pod                  0/1     Completed   0          74m

```

I could even send the requests 
```
curl -v -H 'X-GitHub-Event: pull_request' -H 'X-Hub-Signature: sha1=ba0cdc263b3492a74b601d240c27efe81c4720cb' -H 'Content-Type: application/json' -d '{"action": "opened", "pull_request":{"head":{"sha": "28911bbb5a3e2ea034daf1f6be0a822d50e31e73"}},"repository":{"clone_url": "https://github.com/tektoncd/triggers.git"}}' http://localhost:8080
```
The taskrun created
```
$ kubectl get pods 
NAME                                  READY   STATUS      RESTARTS   AGE
el-github-listener-7ff9675c4d-jpctb   1/1     Running     0          62s
github-run-fhggv-pod                  0/1     Completed   0          74m
github-run-k55bm-pod                  1/1     Running     0          4s
```

**Comment by savitaashture** [comment]
cc @khrm

**Review by khrm — COMMENTED** [review]
Did you test this with OpenShift <=4.15?

**Comment by savitaashture** [comment]
> Did you test this with OpenShift <=4.15?

Yes

**Review by khrm — COMMENTED** [review]
/approve 
@jkandasa

**Review by jkandasa — APPROVED** [review]
cc @piyush-garg

**Comment by jkandasa** [comment]
@savitaashture Please update the release notes

**Comment by piyush-garg** [comment]
i dont know how 65532 is working on openshift, and thing which used to work before is not working now. I need to disucss this more

**Comment by savitaashture** [comment]
> i dont know how 65532 is working on openshift, and thing which used to work before is not working now. I need to disucss this more

Sure @piyush-garg 
should we discuss in Operator WG ?

**Comment by savitaashture** [comment]
When `el-security-context` set to true 
```
 securityContext:
            allowPrivilegeEscalation: false
            # User 65532 is the distroless nonroot user ID
            runAsUser: 65532
            runAsGroup: 65532
            runAsNonRoot: true
            capabilities:
              drop:
                - "ALL"
            seccompProfile:
              type: RuntimeDefault
```
all above things will be set but `65532` is not within the OpenShift range.

Therefore, I need to try out a couple of things by making changes in triggers and operators. 
As a result, we'll be holding this PR for now, and it won't be included in the 1.15 OSP release.

**Review by vdemeester — APPROVED** [review]

File: cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml
Comment (savitaashture) [inline cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: with removal of `runAsUser` and `runAsGroup` **seccompProfiles** is not required for Triggers

But keeping this as it is if at all it require for Pipeline

Comment (piyush-garg) [reply cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: can we check and then do?

Comment (piyush-garg) [reply cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: also it is the pipelines scc, not required for pipeline installation, so dont know if it is required necessarily

Comment (savitaashture) [reply cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: > can we check and then do?

For triggers i did verify and its not required

For pipeline not getting how to ensure may be will create a sample pipelinerun and see if works will remove it

Comment (savitaashture) [reply cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: It doesn't effect to Pipeline when we don't add label

 pod-security.kubernetes.io/enforce=restricted --overwrite=true

But when we add label we get below error for Pipeline
```
$ tkn pr logs -f sum-three-pipeline-run

task first-add has failed: pods "sum-three-pipeline-run-first-add-pod" is forbidden: violates PodSecurity "restricted:latest": unrestricted capabilities (containers "prepare", "step-add" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "prepare", "step-add" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "prepare", "step-add" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
pod for taskrun sum-three-pipeline-run-first-add not available yet
Tasks Completed: 1 (Failed: 1, Cancelled 0), Skipped: 1

```

when we add 
```
seccompProfile:
   type: RuntimeDefault
```

we overcome seccomprofile error but still we have

```
$ tkn pr logs -f sum-three-pipeline-run

task first-add has failed: pods "sum-three-pipeline-run-first-add-pod" is forbidden: violates PodSecurity "restricted:latest": unrestricted capabilities (container "step-add" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "step-add" must set securityContext.runAsNonRoot=true)
pod for taskrun sum-three-pipeline-run-first-add not available yet
Tasks Completed: 1 (Failed: 1, Cancelled 0), Skipped: 1

```

Comment (savitaashture) [reply cmd/openshift/operator/kodata/openshift/00-prereconcile/openshift-pipelines-scc.yaml:36]: Alright as per this https://issues.redhat.com/browse/OCPSTRAT-487 Openshift enforce 
```
pod-security.kubernetes.io/enforce=restricted
```
from 4.17 so we are still good for 4.16

File: pkg/reconciler/openshift/tektontrigger/transformers.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektontrigger/transformers.go:73]: this is already generalized function, please use that

File: components.yaml
Comment (piyush-garg) [inline components.yaml:24]: this is handled by bot

File: pkg/reconciler/kubernetes/tektontrigger/transform.go
Comment (piyush-garg) [inline pkg/reconciler/kubernetes/tektontrigger/transform.go:32]: can we get the key from vendor if exported, otherwise fine

**Comment by piyush-garg** [comment]
please squash the commit into one

File: components.yaml
Comment (savitaashture) [reply components.yaml:24]: Yes 
will update now

File: pkg/reconciler/openshift/tektontrigger/transformers.go
Comment (savitaashture) [reply pkg/reconciler/openshift/tektontrigger/transformers.go:73]: You mean this one https://github.com/tektoncd/operator/blob/main/pkg/reconciler/common/transformers.go#L532 ?

actually i tried that but the function ignore when value is `""`
https://github.com/tektoncd/operator/blob/main/pkg/reconciler/common/transformers.go#L568-L570

and don't want to break existing functionality by doing code changes to that function so written new one considering we need this PR soon so

File: pkg/reconciler/kubernetes/tektontrigger/transform.go
Comment (savitaashture) [reply pkg/reconciler/kubernetes/tektontrigger/transform.go:32]: Its not exposed

But will make a note thank you

File: pkg/reconciler/openshift/tektontrigger/transformers.go
Comment (jkandasa) [reply pkg/reconciler/openshift/tektontrigger/transformers.go:73]: @savitaashture can you try this?
```
import 	"knative.dev/pkg/ptr"

type DefaultRunAsValues struct {
	RunAsUser  *string `json:"default-run-as-user"`
	RunAsGroup *string `json:"default-run-as-group"`
}

emptyDefaultRunAsValues := DefaultRunAsValues{
	RunAsUser:  ptr.String(""),
	RunAsGroup: ptr.String(""),
}
common.AddConfigMapValues(tektontrigger.ConfigDefaults, emptyDefaultRunAsValues)
```

Comment (savitaashture) [reply pkg/reconciler/openshift/tektontrigger/transformers.go:73]: I have updated with existing function with pointer

**Comment by savitaashture** [comment]
> please squash the commit into one

squashed

**Comment by savitaashture** [comment]
@jkandasa @piyush-garg addressed all the review comments
PTAL
Thank you

**Review by jkandasa — APPROVED** [review]
Thanks @savitaashture LGTM
cc @piyush-garg

File: pkg/apis/operator/v1alpha1/tektontrigger_types.go
Comment (piyush-garg) [inline pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: we should not add it here, its a configmap thing and it can be done through options.

Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: we have already stopped adding more configs

Comment (jkandasa) [reply pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: @savitaashture ah I just notice, it is applicable for kubernets too.
Will it work without `runAsUser` and `runAsGroup` in kubernetes?

Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: we should not add it in spec

we just remove those two values in openshift profile only

user should be able to configure through options in both profile

Comment (savitaashture) [reply pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: okay 
got it
then in that case i want to set default value as `""` for both runAsUser and runAsGroup should i add under Options of Triggers  :thinking:

Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektontrigger_types.go:96]: just call the tranformer in openshift extension to set the default value

File: pkg/reconciler/openshift/tektontrigger/extension.go
Comment (jkandasa) [inline pkg/reconciler/openshift/tektontrigger/extension.go:37]: @savitaashture IMO, adding some comments on top of this block will help to understand easily, why we need this change

Comment (savitaashture) [reply pkg/reconciler/openshift/tektontrigger/extension.go:40]: Yes make sense 
added comment

File: pkg/reconciler/openshift/tektontrigger/extension.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektontrigger/extension.go:40]: can we define the struct out of this function

File: pkg/reconciler/openshift/tektontrigger/extension.go
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektontrigger/extension.go:44]: this can also be defined like a constant outside

File: pkg/reconciler/openshift/tektontrigger/extension.go
Comment (savitaashture) [reply pkg/reconciler/openshift/tektontrigger/extension.go]: moved out of the function

**Pull request merged**
